### PR TITLE
feat: vera/proposeEdit enforced edit workflow (#222 Phase F1)

### PR DESCRIPTION
**Pull request by aallan**
## Summary

Phase F1 of the issue 222 skill layer: **`vera/proposeEdit`**, the first enforced workflow method — the whole edit → verify → apply sequence runs server-side as one LSP request, so an agent cannot apply an unverified edit. Applying is the final step of verifying.

Per the Phase F design comment (2026-06-11):

- **New [`vera/lsp/workflows.py`](../blob/feat/222-phase-f1-propose-edit/vera/lsp/workflows.py)** — `propose_edit()` is the pure decision (speculative verify via the Phase E machinery + the gate); `apply_propose_edit()` is the effectful orchestration. The gate: apply iff the proof delta has no `newly_undischarged` obligations AND the proposed state has no error diagnostics; `force: true` overrides both, loudly (the response still carries the full delta).
- **On apply**: a `workspace/applyEdit` server→client request (the client owns the buffer — the server round-trips the edit rather than silently diverging), canonical `DocumentStore` update, re-analysis + diagnostics republish. The client's echoed `didChange` replays from the warm discharge cache — the pre-warming Phase E was designed for.
- **On refuse**: canonical state untouched — same isolation guarantee as `vera/speculativeEdit`.
- **Unopened-URI edge**: full-document replacement uses an exact end position computed from the document's line index (UTF-16 transcoded); for a URI the client never opened, the LSP clamp-sentinel range (clients clamp out-of-range positions to document end).
- Handler registered in `create_server()`; wire glue only.

## Verification gate semantics

Locking: the decision runs under `analysis_lock` (one Z3 session, strictly serialised); the apply path releases before `analyze_and_publish` re-acquires — `threading.Lock` is not reentrant, and the second pass replays from cache by construction.

## Tests (12 new; 4,284 total)

- Gate: clean edit applies; strengthening edit applies (`newly_discharged` must not block); breaking edit refused (violated `nat_sub` surfaced); non-compiling edit refused; force overrides the proof gate AND the error gate with the delta still reported.
- Wiring (structural fake server — transport-free per the established pattern): apply round-trip pins the exact full-document range, store text/version, and republish; refuse touches no canonical state (store, analyses identity, no applyEdit); unopened-URI clamp sentinel.
- `full_document_range` goldens: trailing-newline virtual line; UTF-16 end column on the astral-plane fixture.

## Release prep (v0.0.166)

Version across all 6 tracked sites, CHANGELOG section + compare links, ROADMAP regains a Phase F row (issue 222 is open again and must be tracked; the F3 PR removes it), TESTING.md/README counts, module map (lsp/ 829 → 1007 lines), site assets regenerated.

Part of the Phase F arc (F1 → F2 `strengthenContract` → F3 `addEffect`); this PR intentionally does **not** affect the issue's open state.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added Phase F1: vera/proposeEdit — an enforced edit → verify → apply workflow with gating based on verification results and an explicit force override.

* **Documentation**
  * Bumped project version to 0.0.166; updated changelog, README status, roadmap metrics and module map to reflect the new phase and test counts.

* **Tests**
  * Expanded LSP test coverage with a new suite validating proposeEdit gating, apply behaviour, parameter handling and range/encoding edge cases.

### CODE REVIEW

File: vera/lsp/server.py [hunk 1]
```diff
@@ -161,6 +165,22 @@ def vera_speculative_edit(ls: Any, params: Any) -> dict[str, Any]:
...
+        ``workspace/applyEdit`` + canonical-state update — runs in
+        :func:`vera.lsp.workflows.apply_propose_edit`; this handler is
+        wire glue only.
+        """
+        uri = getattr(params, "uri", None) or params.get("uri")
+        text = getattr(params, "text", None) or params.get("text")
+        force = getattr(params, "force", None)
+        if force is None and hasattr(params, "get"):
+            force = params.get("force")
+        return apply_propose_edit(server, uri, text, bool(force))
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/lsp/server.py]: @coderabbitai Fixed in 4e6db5e — both custom handlers now extract params via a sentinel-based `_param()` helper (missing attribute → dict `.get` → `None`), so falsy-but-present values like `text=""` no longer fall through to `.get` on attribute-style carriers. `TestParamExtraction` pins the `SimpleNamespace(text="")` regression case.

File: tests/test_lsp.py [hunk 2]
```diff
@@ -580,3 +580,198 @@ def test_proof_delta_pure_function(self) -> None:
...
+        store learns the document."""
+        server = _FakeServer()
+        out = apply_propose_edit(server, URI, SPEC_BASE)
+        assert out["applied"] is True
+        (edit,) = server.applied_edits[0].edit.changes[URI]
+        assert edit.range.end.line == 2**31 - 1
+        doc = server.store.get(URI)
+        assert doc is not None
+        assert doc.text == SPEC_BASE
+        assert doc.version == 0
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_lsp.py:797]: @coderabbitai Added `test_client_refusal_does_not_roll_back` in 4e6db5e, but it pins the opposite of the suggested assertions: the workflow is deliberately fire-and-forget on the client's verdict — the response's `applied` reports the gate decision, canonical state is not rolled back, and a declining client re-converges on its next full-sync `didChange` (blocking the handler on the client round-trip would serialise every proposal on editor latency). The fake server now returns a resolved `Future[ApplyWorkspaceEditResult]` mirroring pygls' real signature, so the refusal path is faithfully modelled and the semantics are pinned as documented behaviour in the module docstring.

File: vera/lsp/server.py [hunk 3]
```diff
@@ -161,6 +183,20 @@ def vera_speculative_edit(ls: Any, params: Any) -> dict[str, Any]:
...
+        """#222 Phase F1: enforced edit → verify → apply workflow.
+
+        The whole sequence — speculative verify, gate, and (on pass)
+        ``workspace/applyEdit`` + canonical-state update — runs in
+        :func:`vera.lsp.workflows.apply_propose_edit`; this handler is
+        wire glue only.
+        """
+        uri = _param(params, "uri")
+        text = _param(params, "text")
+        force = bool(_param(params, "force"))
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/lsp/server.py]: @coderabbitai Fixed in 876f54b — `_force_param()` engages only on boolean `true` and fails closed for everything else (strings, numbers, malformed payloads), since force bypasses the verification gate and generic truthiness fails in exactly the wrong direction. `test_force_fails_closed` pins the matrix including the `"false"`-string case you flagged.

File: vera/lsp/server.py [hunk 4]
```diff
@@ -161,6 +196,21 @@ def vera_speculative_edit(ls: Any, params: Any) -> dict[str, Any]:
...
+        The whole sequence — speculative verify, gate, and (on pass)
+        ``workspace/applyEdit`` + canonical-state update — runs in
+        :func:`vera.lsp.workflows.apply_propose_edit`; this handler is
+        wire glue only.
+        """
+        uri = _param(params, "uri")
+        text = _param(params, "text")
+        return apply_propose_edit(
+            server, uri, text, _force_param(params),
+        )
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/lsp/server.py]: @coderabbitai Fixed in a87ac45 — both custom handlers now validate `uri`/`text` at the protocol boundary via `_require_str()`, refusing missing or non-string fields with `JsonRpcInvalidParams` (-32602, the JSON-RPC-native shape) instead of letting `None` surface as an opaque internal error from the parse pipeline. Empty string stays valid (present value — replace-with-empty-document). Tests pin both directions.
